flask/app/genes.py

```diff
@@ -0,0 +1,104 @@
+from csv import DictWriter, QUOTE_MINIMAL
+from io import StringIO
+import traceback
+from flask import abort, jsonify, request, Blueprint, Response, send_file
+from flask_login import login_required
+from sqlalchemy import func
+from .extensions import db
+from .models import Gene
+from .utils import paged
+
+genes_blueprint = Blueprint(
+    "genes",
+    __name__,
+)
+
+
+@genes_blueprint.route("/api/summary/genes", methods=["GET"])
+@paged
+@login_required
+def genes(page, limit):
+    """ index route for genes """
+    search = request.args.get("search", type=str)
+    gene_query = db.session.query(Gene)
+    if search:
+        gene_query = gene_query.filter(func.instr(Gene.hgnc_gene_name, search))
+
+    total_count = gene_query.with_entities(
+        func.count(func.distinct(Gene.gene_id))
+    ).scalar()
+    if limit and page:
+        gene_query.limit(limit).offset(page * limit)
+    results = gene_query.all()
+
+    # default to json if accept = */*
+    if request.accept_mimetypes["application/json"]:
+        return jsonify(
+            {
+                "data": results,
+                "page": page if limit else 0,
+                "total_count": total_count,
+            }
+        )
+
+    if request.accept_mimetypes["text/csv"]:
+        colnames = [col.key for col in Gene.__table__.columns]
+        csv_data = StringIO()
+        writer = DictWriter(
+            csv_data,
+            fieldnames=colnames,
+            quoting=QUOTE_MINIMAL,
+        )
+        writer.writeheader()
+        for row in results:
+            dct = row.__dict__
+            dct.pop("_sa_instance_state")
+            writer.writerow(dct)
+
+        response = Response(csv_data.getvalue())
+
+        response.headers["Content-Disposition"] = "attachment;filename=gene_report.csv"
+        response.headers["Content-Type"] = "text/csv"
+
+        return response
+
+    abort(415, "Mime-Type must be text/csv or application/json!")
+
+
+@genes_blueprint.route("/api/summary/genes/<string:hgnc_gene_name>", methods=["GET"])
+@login_required
+def get_gene_hgnc_name(hgnc_gene_name):
+    """ return a gene based on ghnc gene name """
+    gene_query = db.session.query(Gene)
+    result = gene_query.filter(Gene.hgnc_gene_name == hgnc_gene_name).first()
+
+    if not result:
+        abort(404, f"No gene found with hgnc name {hgnc_name}")
+
+    return jsonify(result)
+
+
+@genes_blueprint.route("/api/summary/genes/hgnc/<int:hgnc_id>", methods=["GET"])
+@login_required
+def get_gene_hgnc_id(hgnc_id):
+    """ return a gene based on ghnc gene id """
+    gene_query = db.session.query(Gene)
+    result = gene_query.filter(Gene.hgnc_gene_id == hgnc_id).first()
+
+    if not result:
+        abort(404, f"No gene found with hgnc_id {hgnc_id}")
+
+    return jsonify(result)
```

Comment:
e.g. in one line since we have no real logging requirement
```suggestion
    return jsonify(Gene.query.filter_by(hgnc_gene_id = hgnc_id).first_or_404())
```

Reply:
updated in [616323a](https://github.com/ccmbioinfo/stager/commit/616323a7b787eef2dd3f960032ee95a1cf415d5f)